File: kotlin/src/main/kotlin/advertisement/application/publishAdvertisement/PublishAdvertisementUseCase.kt
package advertisement.application.publishAdvertisement

import advertisement.domain.AdvertisementRepository
import advertisement.domain.errors.AdvertisementAlreadyExistsError
import advertisement.domain.model.Advertisement
import advertisement.domain.model.value_object.AdvertisementId
import advertisement.domain.model.value_object.Password
import arrow.core.Either
import arrow.core.raise.either
import common.BoundedContextError
import java.time.LocalDateTime

class PublishAdvertisementUseCase(private val advertisementRepository: AdvertisementRepository) {
    fun execute(publishAdvertisementCommand: PublishAdvertisementCommand): Either<BoundedContextError, Unit> {
        return either {
            val advertisementId = AdvertisementId.build(publishAdvertisementCommand.id).bind()
            ensureThatAdvertisementIsUnique(advertisementId).bind()

            val password = Password.fromPlainPassword(publishAdvertisementCommand.password).bind()

            val advertisement = Advertisement.build(
                publishAdvertisementCommand.id,
                publishAdvertisementCommand.description,
                password,
                LocalDateTime.now()
            ).bind()

            advertisementRepository.save(advertisement).bind()
        }
    }

    private fun ensureThatAdvertisementIsUnique(advertisementId: AdvertisementId): Either<BoundedContextError, Unit> {
        return advertisementRepository.findById(advertisementId).fold(
            { Either.Right(Unit) },
            { Either.Left(AdvertisementAlreadyExistsError.withId(advertisementId.value())) } // If Right (success), return Left (error)
        )
    }
}

File: typescript/src/advertisement/application/publish-advertisement/PublishAdvertisementUseCase.ts
import { AdvertisementRepository } from "../../domain/AdvertisementRepository";
import { Advertisement } from "../../domain/model/Advertisement";
import { PublishAdvertisementCommand } from "./PublishAdvertisementCommand";
import {Password} from "../../domain/model/value-object/Password";
import {AdvertisementId} from "../../domain/model/value-object/AdvertisementId";
import {Description} from "../../domain/model/value-object/Description";
import {AdvertisementDate} from "../../domain/model/value-object/AdvertisementDate";
import {sprintf} from "sprintf-js";

export class PublishAdvertisementUseCase {

  constructor(
    private advertisementRepository: AdvertisementRepository
  ) {

  }

  async execute(command: PublishAdvertisementCommand): Promise<void> {
    const advertisementId = new AdvertisementId(command.id)

    if(await this.advertisementRepository.findById(advertisementId)) {
      throw new Error(sprintf('Advertisement with Id %s already exists', advertisementId.value()))
    }

    const advertisement = new Advertisement(
      advertisementId,
      new Description(command.description),
      await Password.fromPlainPassword(command.password),
      new AdvertisementDate(new Date())
    )

    await this.advertisementRepository.save(advertisement)
  }
}
